src/Sylius/Bundle/ResourceBundle/spec/Form/EventListener/AddCodeFormSubscriberSpec.php

```diff
@@ -0,0 +1,81 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace spec\Sylius\Bundle\ResourceBundle\Form\EventListener;
+
+use PhpSpec\ObjectBehavior;
+use Prophecy\Argument;
+use Sylius\Component\Shipping\Model\ShipmentInterface;
+use Sylius\Component\Shipping\Model\ShippingCategoryInterface;
+use Symfony\Component\Form\FormEvent;
+use Symfony\Component\Form\FormEvents;
+use Symfony\Component\Form\FormInterface;
+
+/**
+ * @author Anna Walasek <anna.walasek@lakion.com>
+ */
+class AddCodeFormSubscriberSpec extends ObjectBehavior
+{
+    function it_is_initializable()
+    {
+        $this->shouldHaveType('Sylius\Bundle\ResourceBundle\Form\EventListener\AddCodeFormSubscriber');
+    }
+
+    function it_subscribes_to_event()
+    {
+        $this::getSubscribedEvents()->shouldReturn(array(FormEvents::PRE_SET_DATA => 'preSetData'));
+    }
+
+    function it_sets_code_as_enabled(FormEvent $event, FormInterface $form, ShippingCategoryInterface $shippingCategory)
+    {
+        $event->getData()->willReturn($shippingCategory);
+        $event->getForm()->willReturn($form);
+
+        $shippingCategory->getId()->shouldBeCalled()->willReturn(12);
+
+        $form->add(
+            'code',
+            Argument::type('string'),
+            array('label' => 'sylius.ui.code', 'disabled' => true)
+        )->shouldBeCalled();
+
+        $this->preSetData($event);
+    }
+
+    function it_sets_code_as_disabled(
+        FormEvent $event,
+        FormInterface $form,
+        ShippingCategoryInterface $shippingCategory
+    ) {
+        $event->getData()->willReturn($shippingCategory);
+        $event->getForm()->willReturn($form);
+
+        $shippingCategory->getId()->shouldBeCalled()->willReturn(null);
+
+        $form->add(
+            'code',
+            Argument::type('string'),
+            array('label' => 'sylius.ui.code', 'disabled' => false)
+        )->shouldBeCalled();
+
+        $this->preSetData($event);
+    }
+
+    function it_throws_exception_when_resource_is_not_implement_code_aware_interface(FormEvent $event, ShipmentInterface $shipment)
+    {
+        $event->getData()->willReturn($shipment);
```

Comment:
I think better would be to use `new \stdClass()` since maybe some day we will implement `CodeAwareInterface` on Shipping also :)